### PR TITLE
DRILL-6870: Upgrade to ANTLR4

### PR DESCRIPTION
**Pull request by arina-ielchiieva**
Link to Jira -  [DRILL-6870](https://issues.apache.org/jira/browse/DRILL-6870).

### CODE REVIEW

File: logical/src/main/antlr4/org/apache/drill/common/expression/parser/ExprParser.g4 [hunk 1]
```diff
@@ -55,25 +47,20 @@ import org.apache.drill.common.exceptions.ExpressionParsingException;
...
+    if (fullExpression == null) fullExpression = $expression.text;
     tokenPos = $expression.start.getTokenIndex();
   }
   ;
  
 functionCall returns [LogicalExpression e]
-  :  Identifier OParen exprList? CParen {$e = FunctionCallFactory.createExpression($Identifier.text, pos($Identifier), $exprList.listE);  }
+  :  Identifier OParen exprList? CParen {$e =
+      FunctionCallFactory.createExpression($Identifier.text, pos($Identifier),
+        ($exprList.ctx == null ? new ArrayList<LogicalExpression>() : $exprList.listE)); }
```
Comment (vvysotskyi) [inline logical/src/main/antlr4/org/apache/drill/common/expression/parser/ExprParser.g4]: Is diamond syntax supported here? If yes, could you please remove type here?

File: logical/src/main/java/org/apache/drill/common/parser/ErrorListener.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  public static final ErrorListener INSTANCE = new ErrorListener();
+
+  @Override
+  public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line,
+                          int charPositionInLine, String msg, RecognitionException e) {
+    throw new ExpressionParsingException(msg);
+  }
+
+}
```
Comment (vvysotskyi) [inline logical/src/main/java/org/apache/drill/common/parser/ErrorListener.java]: Could you please add a new line here?

File: logical/pom.xml [hunk 3]
```diff
@@ -112,15 +111,12 @@
       </plugin>
       <plugin>
         <groupId>org.antlr</groupId>
-        <artifactId>antlr3-maven-plugin</artifactId>
-        <version>3.4</version>
-        <configuration>
-
-        </configuration>
+        <artifactId>antlr4-maven-plugin</artifactId>
+        <version>4.7.1</version>
```
Comment (vvysotskyi) [inline logical/pom.xml]: Can be used `${antlr.version}` here?

File: logical/src/main/antlr4/org/apache/drill/common/expression/parser/ExprParser.g4 (hunk 1 above)
Comment (arina-ielchiieva) [reply logical/src/main/antlr4/org/apache/drill/common/expression/parser/ExprParser.g4]: Agree, removed in other places as well.

File: logical/src/main/java/org/apache/drill/common/parser/ErrorListener.java (hunk 2 above)
Comment (arina-ielchiieva) [reply logical/src/main/java/org/apache/drill/common/parser/ErrorListener.java]: Done.

File: logical/pom.xml (hunk 3 above)
Comment (arina-ielchiieva) [reply logical/pom.xml]: Sure, I'll make the change.

**Comment by arina-ielchiieva** [comment]
@vvysotskyi addressed code review comments, thanks for the review!

**Pull request closed without merging**